### PR TITLE
[ENH] Allow object dtype in series

### PR DESCRIPTION
**Pull request by yarnabrina**
Reference: https://github.com/sktime/sktime/issues/5867#issuecomment-1917438371

### CODE REVIEW

**Comment by yarnabrina** [comment]
@fkiraly I see a new failure already, but I don't know what part of my change is affecting this. Can you guide?

Also, I have a question. What is the difference between `mtype` and `scitype`?

**Comment by fkiraly** [comment]
scitype = [abstract data type](https://en.wikipedia.org/wiki/Abstract_data_type), e.g., time series, or collection of time series
mtype = full python specification

**Comment by fkiraly** [comment]
The failures are due to a secondary problem, namely that objects of mtype `nested_univ` now also pass the check for `pd.DataFrame`, in addition to the `nested_univ` check.

That means the two mtypes are no longer distinguishable by checks, which leads to further trouble as the data checking layer can no longer decide what mtype it shoud assume

I am guessing that this is because the `nested_univ` check is too permissive, e.g., does not check that the entries indeed contain `pd.Series`. A fix would be to look at that check, and perhaps check that at east one of the entries in the first row contains a `pd.Series` typed object

**Comment by fkiraly** [comment]
should I try to do that?

**Comment by yarnabrina** [comment]
I'm unlikely to any time this week, so please do if you have time. I'll
attempt over weekend otherwise.

**Comment by fkiraly** [comment]
@yarnabrina, your remote is rejecting push. You need to check the box "allow maintainers to make changes".

To fix, add following lines starting line 66 of `_check._series`:

```python
    # check to delineate from nested_univ mtype (Panel)
    # pd.DataFrame mtype allows object dtype,
    # but if we allow object dytpe with pd.Series entries,
    # the mtype becomes ambiguous, i.e., non-delineable from nested_univ
    if isinstance(obj.iloc[0, 0], (pd.Series, pd.DataFrame)):
        msg = f"{var_name} cannot contain nested pd.Series or pd.DataFrame"
        return ret(False, msg, None, return_metadata)
```

**Comment by yarnabrina** [comment]
> You need to check the box "allow maintainers to make changes".

Done, can you please check now?

**Comment by fkiraly** [comment]
ok, works

**Comment by yarnabrina** [comment]
@fkiraly I'll try to get back to this PR on the weekend. Can you give some suggestions about the new errors? I definitely did not expect IndexError to come after allowing a new dtype.

**Comment by fkiraly** [comment]
> Can you give some suggestions about the new errors? I definitely did not expect IndexError to come after allowing a new dtype.

tried to fix - failure is incomplete edge case treatment of empty data frames (zero rows or cols)

**Comment by yarnabrina** [comment]
@fkiraly making it as ready, all tests pass after that additional check you added.

**Review by fkiraly — APPROVED** [review]
So, works now and I'd be happy with it

The change I made was the additional check in now lines 66-72 of `_series._check`, which allows the system to delineate `nested_univ` from `pd.DataFrame` mtype, at least in all test cases. Empty `pd.DataFrame` count as the latter.

**Comment by yarnabrina** [comment]
> Panel and Hierarchical types still forbid object

I actually skipped this one intentionally. I thought it may be better to allow separately for different types. Unless you think they are interlinked and can cause issues, I really have very little, if any, familiarity with how scitype/mtype/etc. works in sktime?

> many estimators will break if given object dtype-d data

This one is sort of a dilemma - currently a lot can use object type and we don't let them use it, while allowing may break some which do support. I don't know how to choose. Also, there's a "category" type which is being more and more popular. `pandas` has recently planning to use even a dedicated "string" type, I think I saw in their latest release notes.

> we are not systematically testing this input type - from now, users will not receive clear warning.

I didn't understnad what you meant by this one. What is happening now, and what will change to confuse users?

**Comment by fkiraly** [comment]
> Unless you think they are interlinked and can cause issues

Yes, they are linked through broadcasting.

One potential failure case that is not covered due to lack of examples for the contract expansion is as follows:

suppose we have a transformer that internally is implemented only for `Panel` (e.g., for efficiency reasons).
Upon receiving a `Series`, it will be upcast to `Panel`, and suddenly the object no longer satisfies the contract.

Similarly, there is now an inconsistency the other way round - even if an estimator would now accept an individual series, it would no longer accept it if it is found with others of similar type in a collection of series.

PS: I think for `Panel`, the trick with the `iloc` is not needed, as the delineation happens via the `MultiIndex` already.

**Comment by fkiraly** [comment]
> This one is sort of a dilemma - currently a lot can use object type and we don't let them use it, while allowing may break some which do support. I don't know how to choose. Also, there's a "category" type which is being more and more popular

I do think ew should support it, but "properly".
Here's what I suggest:

* extend the types for `Panel` and `Hierarchical` as well
* in `scenarios_forecasting` and `scenarios_transformers` add coverage for this
    * if not too much breaks, hooray - fix the few bugs etc
    * if a lot breaks, we need to consider how we make the contract tight. A way would be introducing tags for the supported types.

**Comment by fkiraly** [comment]
>> we are not systematically testing this input type - from now, users will not receive clear warning.

> I didn't understnad what you meant by this one. What is happening now, and what will change to confuse users?

Consider the situation of an estimator that does not support the type genuinely, i.e., would break internally of `pd.DataFrame` with object `dtype` is passed.

Suppose a user passes a `pd.DataFrame` with object `dtype` to that estimator. Currently, on `main`, this will fail with an informative error message, namely "object `dtype` is not supported" or similar. This is because the check catches it, and also returns an informative error message together with the check.

However, if we widen the contract without testing and adding catches where they fail, then in this estimator something random will break, and the error will be uninformative, e.g., a `ValueError` or type error or similar. This is of course not good.

From a higher poin of view, by widening the contract, we may now be creating uncovered bugs. So either we have to move the contract boundary "further in" by one of the measures proposed above (e.g., tag and catch), or we have to ensure all estimators satisfy the new contract. In any case, it has to be tested.

**Comment by fkiraly** [comment]
oh wait, the problem is actually `Panel`. I am reverting the change I made to the `Series` checks.

**Comment by fkiraly** [comment]
to avoid conflicts: do you want ot add the mini-check in `Panel`, or should I?

**Comment by yarnabrina** [comment]
Please do. I've very little experience and confidence in this datatypes modules.

File: sktime/datatypes/_panel/_check.py
Comment (yarnabrina) [inline sktime/datatypes/_panel/_check.py:196]: I just copy pasted your comment and tried to edit two terms, but I'm certain this is wrong. So, please review and change.

**Review by yarnabrina — COMMENTED** [review]
I did a rebase from main and force push, so I think you will get conflicts. So, if possible can you delete your local branch and pull this?

Also, I do not know where are the other changes you said are necessary to do, so can you please apply that? I'll not modify this branch anymore to avoid future conflicts, deleting from my local.

File: sktime/datatypes/_panel/_check.py
Comment (fkiraly) [reply sktime/datatypes/_panel/_check.py:196]: that should work though, let's see what the tests say

**Comment by yarnabrina** [comment]
FYI @fkiraly @Abhay-Lejith 

I updated this branch with `main` branch (with rebase), but cancelled the CI to avoid affecting other release PR's. Also, converting this into draft as I believe we're going to wait for the series of PR plan Abhay listed in the umbrella issue to be ready.

**Comment by yarnabrina** [comment]
@fkiraly I merged #6704 and #6490 as per the discussion in Discord and #6109, and updated this PR with `main`. Assuming tests pass (no unexpected failure in @Abhay-Lejith's fork's test-all), can you please review/approve/merge this PR to unblock the next PR in sequence, viz. #6732?

**Review by fkiraly — APPROVED** [review]
sure!

**Pull request merged**
